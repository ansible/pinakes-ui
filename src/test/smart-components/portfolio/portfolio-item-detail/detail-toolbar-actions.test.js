import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';

import DetailToolbarActions from '../../../../smart-components/portfolio/portfolio-item-detail/detail-toolbar-actions';
import { Dropdown } from '@patternfly/react-core';

const prepareTruthyCapability = (truthyCapability) => ({
  copy: false,
  update: false,
  tags: false,
  ...(truthyCapability
    ? {
        [truthyCapability]: true
      }
    : {})
});

describe('<DetailToolbarActions />', () => {
  let initialProps;

  beforeEach(() => {
    initialProps = {
      orderUrl: 'foo/bar',
      editUrl: 'foo/baz',
      isOpen: false,
      setOpen: jest.fn(),
      copyUrl: 'foo/copy',
      availability: 'available',
      editSurveyUrl: 'foo/edit-survey',
      workflowUrl: 'foo/workflow',
      pathname: 'foo/bar',
      userCapabilities: {
        copy: true,
        update: true,
        tags: true
      },
      canLinkOrderProcesses: true
    };
  });

  it('should render correctly', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions {...initialProps} />
      </MemoryRouter>
    );
    expect(toJson(wrapper.find(DetailToolbarActions))).toMatchSnapshot();
  });

  it('should show copy action in dropdown', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions
          {...initialProps}
          isOpen
          userCapabilities={prepareTruthyCapability('copy')}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('li')).toHaveLength(1);
    expect(wrapper.find('li#copy-portfolio-item')).toHaveLength(1);
  });

  it('should disable the Order button if orderable is false', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions
          {...initialProps}
          isOpen
          orderable={false}
          userCapabilities={prepareTruthyCapability('copy')}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('Button').props().isDisabled).toBeTruthy();
  });

  it('should disable the Order button if orderable is false', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions
          {...initialProps}
          isOpen
          orderable={true}
          userCapabilities={prepareTruthyCapability('copy')}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('Button').props().isDisabled).toBeFalsy();
  });

  it('should show edit, set order processes and edit survey actions in dropdown', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions
          {...initialProps}
          isOpen
          userCapabilities={prepareTruthyCapability('update')}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('li')).toHaveLength(3);
    expect(wrapper.find('li#edit-portfolio-item')).toHaveLength(1);
    expect(wrapper.find('li#edit-survey')).toHaveLength(1);
    expect(wrapper.find('li#attach-order-processes')).toHaveLength(1);
  });

  it('should show set_approval action in dropdown', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions
          {...initialProps}
          isOpen
          userCapabilities={prepareTruthyCapability('tags')}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('li')).toHaveLength(1);
    expect(wrapper.find('li#set-approval_workflow')).toHaveLength(1);
  });

  it('should not render dropdown', () => {
    const wrapper = mount(
      <MemoryRouter>
        <DetailToolbarActions {...initialProps} isOpen userCapabilities={{}} />
      </MemoryRouter>
    );

    expect(wrapper.find(Dropdown)).toHaveLength(0);
  });
});
