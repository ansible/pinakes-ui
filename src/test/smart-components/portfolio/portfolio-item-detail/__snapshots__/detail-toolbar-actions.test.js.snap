// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailToolbarActions /> should render correctly 1`] = `
<DetailToolbarActions
  availability="available"
  canLinkOrderProcesses={true}
  copyUrl="foo/copy"
  editSurveyUrl="foo/edit-survey"
  editUrl="foo/baz"
  isFetching={false}
  isOpen={false}
  orderUrl="foo/bar"
  orderable={true}
  pathname="foo/bar"
  setOpen={[MockFunction]}
  userCapabilities={
    Object {
      "copy": true,
      "tags": true,
      "update": true,
    }
  }
  workflowUrl="foo/workflow"
>
  <LevelItem>
    <div>
      <CatalogLink
        isDisabled={false}
        pathname="foo/bar"
        preserveSearch={true}
      >
        <Styled(Component)
          isDisabled={false}
          to={
            Object {
              "hash": undefined,
              "pathname": "foo/bar",
              "search": "",
            }
          }
        >
          <Component
            className="sc-AxiKw drCTeX"
            isDisabled={false}
            to={
              Object {
                "hash": undefined,
                "pathname": "foo/bar",
                "search": "",
              }
            }
          >
            <Link
              className="sc-AxiKw drCTeX"
              to={
                Object {
                  "hash": undefined,
                  "pathname": "foo/bar",
                  "search": "",
                }
              }
            >
              <LinkAnchor
                className="sc-AxiKw drCTeX"
                href="foo/bar"
                navigate={[Function]}
              >
                <a
                  className="sc-AxiKw drCTeX"
                  href="foo/bar"
                  onClick={[Function]}
                >
                  <ButtonWithSpinner
                    id="order-portfolio-item"
                    isDisabled={false}
                    ouiaId="order-portfolio-item"
                    showSpinner={false}
                    variant="primary"
                  >
                    <Styled(Button)
                      id="order-portfolio-item"
                      isDisabled={false}
                      ouiaId="order-portfolio-item"
                      variant="primary"
                    >
                      <Button
                        className="sc-AxjAm jhCAUI"
                        id="order-portfolio-item"
                        isDisabled={false}
                        ouiaId="order-portfolio-item"
                        variant="primary"
                      >
                        <ButtonBase
                          className="sc-AxjAm jhCAUI"
                          id="order-portfolio-item"
                          innerRef={null}
                          isDisabled={false}
                          ouiaId="order-portfolio-item"
                          variant="primary"
                        >
                          <button
                            aria-disabled={false}
                            aria-label={null}
                            className="pf-c-button pf-m-primary sc-AxjAm jhCAUI"
                            data-ouia-component-id="order-portfolio-item"
                            data-ouia-component-type="PF4/Button"
                            data-ouia-safe={true}
                            disabled={false}
                            id="order-portfolio-item"
                            role={null}
                            type="button"
                          >
                            <styled.span
                              showSpinner={false}
                            >
                              <span
                                className="sc-AxirZ fnlrkz"
                              >
                                Order
                              </span>
                            </styled.span>
                          </button>
                        </ButtonBase>
                      </Button>
                    </Styled(Button)>
                  </ButtonWithSpinner>
                </a>
              </LinkAnchor>
            </Link>
          </Component>
        </Styled(Component)>
      </CatalogLink>
    </div>
  </LevelItem>
  <LevelItem
    style={
      Object {
        "marginLeft": 16,
      }
    }
  >
    <div
      style={
        Object {
          "marginLeft": 16,
        }
      }
    >
      <Dropdown
        dropdownItems={
          Array [
            <DropdownItem
              aria-label="Edit Portfolio"
              component={
                <CatalogLink
                  pathname="foo/baz"
                  preserveSearch={true}
                >
                  Edit
                </CatalogLink>
              }
              id="edit-portfolio-item"
              role="link"
            />,
            <DropdownItem
              aria-label="Copy Portfolio"
              component={
                <CatalogLink
                  pathname="foo/copy"
                  preserveSearch={true}
                >
                  Copy
                </CatalogLink>
              }
              id="copy-portfolio-item"
              role="link"
            />,
            <DropdownItem
              aria-label="Set approval"
              component={
                <CatalogLink
                  pathname="foo/workflow"
                  preserveSearch={true}
                >
                  Set approval
                </CatalogLink>
              }
              id="set-approval_workflow"
              role="link"
            />,
            <DropdownItem
              aria-label="Set order processes"
              component={
                <CatalogLink
                  pathname="/portfolios/portfolio/portfolio-item/edit-order-process"
                  preserveSearch={true}
                >
                  Set order processes
                </CatalogLink>
              }
              id="attach-order-processes"
              role="link"
            />,
            <DropdownItem
              aria-label="Edit survey"
              component={
                <CatalogLink
                  pathname="foo/edit-survey"
                  preserveSearch={true}
                >
                  Edit survey
                </CatalogLink>
              }
              id="edit-survey"
              role="link"
            />,
          ]
        }
        isOpen={false}
        isPlain={true}
        onSelect={[Function]}
        onToggle={[MockFunction]}
        position="right"
        toggle={
          <KebabToggle
            id="portfolio-item-actions-toggle"
            onToggle={[Function]}
          />
        }
      >
        <DropdownWithContext
          autoFocus={true}
          className=""
          direction="down"
          dropdownItems={
            Array [
              <DropdownItem
                aria-label="Edit Portfolio"
                component={
                  <CatalogLink
                    pathname="foo/baz"
                    preserveSearch={true}
                  >
                    Edit
                  </CatalogLink>
                }
                id="edit-portfolio-item"
                role="link"
              />,
              <DropdownItem
                aria-label="Copy Portfolio"
                component={
                  <CatalogLink
                    pathname="foo/copy"
                    preserveSearch={true}
                  >
                    Copy
                  </CatalogLink>
                }
                id="copy-portfolio-item"
                role="link"
              />,
              <DropdownItem
                aria-label="Set approval"
                component={
                  <CatalogLink
                    pathname="foo/workflow"
                    preserveSearch={true}
                  >
                    Set approval
                  </CatalogLink>
                }
                id="set-approval_workflow"
                role="link"
              />,
              <DropdownItem
                aria-label="Set order processes"
                component={
                  <CatalogLink
                    pathname="/portfolios/portfolio/portfolio-item/edit-order-process"
                    preserveSearch={true}
                  >
                    Set order processes
                  </CatalogLink>
                }
                id="attach-order-processes"
                role="link"
              />,
              <DropdownItem
                aria-label="Edit survey"
                component={
                  <CatalogLink
                    pathname="foo/edit-survey"
                    preserveSearch={true}
                  >
                    Edit survey
                  </CatalogLink>
                }
                id="edit-survey"
                role="link"
              />,
            ]
          }
          isGrouped={false}
          isOpen={false}
          isPlain={true}
          isText={false}
          menuAppendTo="inline"
          onSelect={[Function]}
          onToggle={[MockFunction]}
          position="right"
          toggle={
            <KebabToggle
              id="portfolio-item-actions-toggle"
              onToggle={[Function]}
            />
          }
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-1"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
            onToggle={[MockFunction]}
          >
            <KebabToggle
              aria-haspopup={true}
              getMenuRef={[Function]}
              id="portfolio-item-actions-toggle"
              isOpen={false}
              isPlain={true}
              isText={false}
              key=".0"
              onEnter={[Function]}
              onToggle={[Function]}
              parentRef={
                Object {
                  "current": <div
                    class="pf-c-dropdown pf-m-align-right"
                    data-ouia-component-id="OUIA-Generated-Dropdown-1"
                    data-ouia-component-type="PF4/Dropdown"
                    data-ouia-safe="true"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-label="Actions"
                      class="pf-c-dropdown__toggle pf-m-plain"
                      id="portfolio-item-actions-toggle"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 192 512"
                        width="1em"
                      >
                        <path
                          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                        />
                      </svg>
                    </button>
                  </div>,
                }
              }
            >
              <Toggle
                aria-haspopup={true}
                aria-label="Actions"
                bubbleEvent={false}
                className=""
                getMenuRef={[Function]}
                id="portfolio-item-actions-toggle"
                isActive={false}
                isDisabled={false}
                isOpen={false}
                isPlain={true}
                isPrimary={false}
                isSplitButton={false}
                isText={false}
                onEnter={[Function]}
                onToggle={[Function]}
                parentRef={
                  Object {
                    "current": <div
                      class="pf-c-dropdown pf-m-align-right"
                      data-ouia-component-id="OUIA-Generated-Dropdown-1"
                      data-ouia-component-type="PF4/Dropdown"
                      data-ouia-safe="true"
                    >
                      <button
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-label="Actions"
                        class="pf-c-dropdown__toggle pf-m-plain"
                        id="portfolio-item-actions-toggle"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 192 512"
                          width="1em"
                        >
                          <path
                            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          />
                        </svg>
                      </button>
                    </div>,
                  }
                }
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-label="Actions"
                  className="pf-c-dropdown__toggle pf-m-plain"
                  disabled={false}
                  id="portfolio-item-actions-toggle"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <EllipsisVIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style={
                        Object {
                          "verticalAlign": "-0.125em",
                        }
                      }
                      viewBox="0 0 192 512"
                      width="1em"
                    >
                      <path
                        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                      />
                    </svg>
                  </EllipsisVIcon>
                </button>
              </Toggle>
            </KebabToggle>
          </div>
        </DropdownWithContext>
      </Dropdown>
    </div>
  </LevelItem>
</DetailToolbarActions>
`;
